File: frontend-app/src/app/services/tipo-ahorro.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { TipoAhorro } from "../domain/TipoAhorro";


@Injectable({
    providedIn: 'root'
})
export class TipoAhorroService{

    private url:string = 'http://localhost:8080/api/tipos-ahorros';

    constructor(private http: HttpClient){

    }

    listadoTiposAhorros(): Observable<TipoAhorro[]>{
        return this.http.get<TipoAhorro[]>(this.url);
    }
    
        
    buscarTipoAhorroPorId(id: number): Observable<TipoAhorro>{
        return this.http.get<TipoAhorro>(`${this.url}/${id}`);
    }

    guardarTipoAhorro(tipoAhorro: TipoAhorro){
        return this.http.post('http://localhost:8080/api/tipos-ahorros/guardar', tipoAhorro);
    }
    
    editarTipoAhorro(tipoAhorro: TipoAhorro) {
        return this.http.put(`${this.url}/actualizar`, tipoAhorro);
    }
    
    eliminarTipoAhorro(idTipoAhorro: number){
        return this.http.delete(`${this.url}/${idTipoAhorro}`);
    }    

    listadoTipoAhorroPaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }

    buscarTipoAhorroPorNombre(nombreTipoAhorro: string): Observable<TipoAhorro[]> {
        const params = new HttpParams().set('nombreTipoAhorro', nombreTipoAhorro);
    
        return this.http.get<TipoAhorro[]>(`${this.url}/buscar`, { params });
    }

    listadoTiposAhorrosCompletos(): Observable<TipoAhorro[]>{
        return this.http.get<TipoAhorro[]>(`${this.url}/completos`);
    }

}

File: backend-app/src/main/java/com/backend_app/backend_app/service/TipoAhorroService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.TipoAhorro;
import com.backend_app.backend_app.dto.TipoAhorroDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface TipoAhorroService {

    List<TipoAhorroDTO> listadoTiposAhorro();

    Optional<TipoAhorroDTO> buscarTipoAhorroPorId(long idTipoAhorro);

    TipoAhorroDTO convertirADTO(TipoAhorro tipoAhorro);

    void insertarTipoAhorro(TipoAhorroDTO tipoAhorro);

    void actualizarTipoAhorro(TipoAhorroDTO tipoAhorro);

    void eliminarTipoAhorro(TipoAhorroDTO tipoAhorro);

    Page<TipoAhorroDTO> findAll(Pageable pageable);

    List<TipoAhorroDTO> buscarTipoAhorroPorNombre(String nombreTipoAhorro);

    List<TipoAhorroDTO> listadoTiposAhorrosCompletos();

}
